### PR TITLE
feat(rootfs): native armhf on aarch64 host via runtime-disable of qemu-arm

### PR DESCRIPTION
**Pull request by iav**
> **Fork-internal codex review trap.** Mirrors armbian/build#9769 (\`feat/native-armhf-on-arm64\`, head `5cab79ef`). Not for merge. CodeRabbit reviews on armbian/build#9769 (paid by Armbian, only sees upstream PRs); Codex reviews here (paid by iav, only sees fork PRs). Re-trigger codex on each force-push via `@chatgpt-codex-connector review`.
>
> --- Below: verbatim copy of the upstream PR description for codex context. ---

## Summary

Continues #9284 (`arm64-compat-vdso` extension + `custom_kernel_make_params` hook).
That PR enabled `COMPAT_VDSO` in arm64 kernels; this PR uses the resulting
32-bit-capable kernels to run 32-bit ARM ELF directly instead of through
`qemu-arm` emulation in chroot work — typically ~12× faster than `qemu-arm-static`
on Cortex-A53/A55/A72/A73/A76.

## How

Disable `qemu-arm` in `/proc/sys/fs/binfmt_misc/` for the duration of the
build; kernel `binfmt_elf` then handles 32-bit ARM ELFs natively via
`CONFIG_COMPAT`. Activation is delayed until AFTER `mmdebstrap`, whose
cross-arch path needs a working `qemu-arm` registration to populate the
chroot before `libc/ld-linux-armhf.so.3` exist.

Killswitch: `NATIVE_ARMHF_ON_ARM64=no` (synonyms `never`, `disabled`).

## Concurrency — kernel `flock` on the binfmt entry

Concurrent armbian builds on the same host kernel are coordinated purely
via kernel BSD `flock(2)` on `/proc/sys/fs/binfmt_misc/qemu-arm` itself.
No userspace state, no per-builder lockfiles, no `/run/lock` directory.

- Each builder holds `LOCK_SH` on the binfmt entry via a long-lived fd.
  Kernel BSD-flock counter is the refcount; kernel releases the fd on
  process exit (crash-safe).
- First-arrival idempotently `echo 0` to disable; subsequent arrivals
  observe 0 and proceed without writing.
- On exit, release `LOCK_SH`; last-out detects via `LOCK_EX-LOCK_NB` on
  a fresh fd, succeeds iff zero other `LOCK_SH` holders. Last-out
  re-enables `qemu-arm`.

## Trade-offs

1. Prior `qemu-arm` state is not recorded across builds. Last-out
   unconditionally re-enables. An admin's pre-existing `disabled`
   policy is not preserved across the build window.
2. No defense against an external agent toggling `qemu-arm` mid-build.
   Don't, or use `NATIVE_ARMHF_ON_ARM64=no` on builds that should not
   interfere.
3. Cleanup ordering invariant — BSD flock is per-OFD; a forked subshell
   inheriting our SH-fd shares the same lock entry. We rely on
   `add_cleanup_handler` running registrations in order and the umount
   teardown handlers (registered first) killing the docker container
   before our restore handler runs. Documented in the trap handler's
   docstring; if a future refactor breaks ordering, POSIX `F_SETLK` or
   explicit descendant-kill are the documented escape hatches.

The earlier revision of this work coordinated via a userspace
owner-flock directory (~250 lines, depended on `/run/lock` being
host-bind-mounted into build containers — which Armbian docker does
not do). Replaced with the kernel-flock approach above; available in
PR history if needed.

## Empirical

`./compile.sh build BOARD=helios4 BRANCH=edge BUILD_MINIMAL=yes RELEASE=noble`
on droid (RK3328) and m1 (RK3568, Cortex-A55):

| Scenario | Real time | Speedup |
|---|---|---|
| baseline, rootfs cache-miss | 60:35 | 1× |
| this PR, rootfs cache-miss  | 19:27 | 3.12× |
| this PR, rootfs cache-hit   | 6:38  | ~9× over expected baseline cache-hit |

`mmdebstrap` itself (~10 min of every cache-miss build) cannot be
accelerated — speedup applies to subsequent chroot operations.

## Test plan

Phase 1 — feature on iron (single builder):

- [x] helios4 cache-miss + cache-hit, full image
- [x] cleanup re-enables `qemu-arm` on exit
- [x] `DOCKER_PRIVILEGED=no` (CAP_SYS_ADMIN unconditional)
- [x] killswitch (`NATIVE_ARMHF_ON_ARM64=no/never/disabled`) — silent return
- [x] x86_64 host regression — silent return
- [x] aarch64 native (no-container) `PREFER_DOCKER=no` smoke

Phase 2 — concurrency, all on Hetzner CAX21 (kernel 6.8.0-90, binfmt_misc fs):

- [x] flock primitives on `/proc/sys/fs/binfmt_misc/qemu-arm` work as expected
- [x] single-tree build: setup + 5 phase markers + last-out restore
- [x] two independent worktrees concurrent: both enter native path, kernel
      SH-counter holds qemu-arm disabled, first-finished suppresses restore,
      last-finished restores
- [x] `NATIVE_ARMHF_ON_ARM64=no` opt-out with concurrent native builder —
      refused-fast guard rejects (avoids clobbering)
- [x] SIGINT mid-chroot — cleanup ordering invariant holds; last-out fires,
      qemu-arm restored, no stale state

Reviewed with codex (gpt-5.5) over multiple rounds and coderabbitai. The
final review surfaced the cleanup-ordering invariant; documented inline
and verified by SIGINT test above.

## Summary by CodeRabbit

* **New Features**
  - Enable native 32-bit ARM (armhf) execution on ARM64 hosts during builds so chroot operations can run 32-bit ARM binaries without injecting emulation binaries.

* **Chores**
  - Coordinate host binfmt/qemu management directly via kernel `flock` on `/proc/sys/fs/binfmt_misc/qemu-arm`; first-arrival disables qemu-arm, last-departure restores it. Setup failures are non‑fatal so image builds continue.

## Summary by CodeRabbit

* **New Features**
  * Allow native execution of 32-bit ARM binaries on compatible 64-bit hosts to speed and simplify ARM image builds.
  * Skip copying emulation binaries into chroots when native native-ARM handling is active.

* **Bug Fixes**
  * Improve coordination and recovery around host binary-format handling to avoid interference between concurrent builds and ignore non-fatal setup failures.

[Review Change Stack](https://app.coderabbit.ai/change-stack/armbian/build/pull/9769)

### CODE REVIEW

File: lib/functions/rootfs/qemu-static.sh [hunk 1]
```diff
@@ -201,12 +427,13 @@ function prepare_host_binfmt_qemu_cross_arm64_host_armhf_target() {
 		run_host_command_logged arch-test "||" true
 	fi
 
-	# to check, we use arch-test; if will return 0 if _either_ the host can natively run armhf, or if qemu-arm is correctly working.
-	if arch-test arm; then
+	# to check, we use arch-test; will return 0 if _either_ the host can natively run armhf, or if qemu-arm is correctly working.
+	# Use armhf (Debian-arch) rather than arm to match the build target and the post-disable check in _native_armhf_setup_binfmt_elf.
+	if arch-test armhf; then
```
[the remark this answers is not in the document]
Comment (iav) [reply lib/functions/rootfs/qemu-static.sh]: The killswitch (`NATIVE_ARMHF_ON_ARM64=no`) prevents *this builder* from modifying `/proc/sys/fs/binfmt_misc/qemu-arm` — its purpose is binfmt_misc non-interference with concurrent builds, not enforcing a specific execution mode.

On a native-capable aarch64 host where qemu-arm is absent: `arch-test armhf` succeeds because the kernel handles armhf ELFs natively. That is correct behavior — there is nothing to fall back to. If the operator wants to force qemu emulation on such a host, installing and registering qemu-arm is a prerequisite; the killswitch does not and should not paper over a missing qemu installation.

Not a bug in scope for this PR.

**Comment by iav** [comment]
@chatgpt-codex-connector review

New commit `ecde3c1` addresses:

- **Race in N-builder entry (line:239)** — restructured to hold EX during state-check + `echo 0`, then atomically downgrade EX→SH on the same fd. Linux `flock(2)` performs the transition under the inode's `flc_lock`, so SH waiters are granted only after our SH is in place and `qemu-arm` is already 0.

- **K-builder symmetric check** — after acquiring the emulation-mode SH-NB, K verifies `state=1`; if 0 (concurrent N-builders already disabled qemu-arm), bail with a clear error instead of running a broken anchor.

- **Cleanup-ordering invariant comment** — was inverted. `add_cleanup_handler` PREPENDS, so later registrations run FIRST. The invariant survives because setup is invoked before `mount_chroot` in both call sites; the new comment states the actual mechanism and the failure mode if a future call site inverts the order.

`line:24` (preserve target qemu binaries) and `line:432` (killswitch + qemu absent) — deferred / not-a-bug; replies in their threads.

**Comment by iav** [comment]
@chatgpt-codex-connector review

Force-pushed `6138bd03` addressing all three P1/P2 findings from your previous review:

- **P1 (line:407 → removed)** — early `_native_armhf_setup_binfmt_elf` call in `prepare_host_binfmt_qemu_cross` deleted. Native activation stays where it belongs: post-mmdebstrap (`rootfs-create.sh`, `rootfs-image.sh`). The container relaunch / Apple-Silicon race the early call was trying to mitigate is a separate concern — addressing it without breaking cache-miss mmdebstrap is out of scope here.

- **P2 (line:275 → joiner state-recheck)** — added in the `else` branch. After `flock -s -w 30` succeeds, observe state again; if the last-out restorer raced in and wrote `qemu-arm` back to 1, release SH and fall back to emulation rather than continuing as a phantom native joiner.

- **P2 (line:189 → killswitch ARCH-gated)** — moved idempotent + ARCH/host gates above killswitch evaluation. Unrelated builds (amd64/arm64 targets, x86 hosts) skip the qemu-arm SH-anchor entirely instead of pointlessly holding it.

Build cache-miss test on real aarch64 host (m1, RK3568) with the prior `ecde3c1` is still running in the background — will report empirical confirmation once it completes mmdebstrap.

**Comment by iav** [comment]
@chatgpt-codex-connector review

Force-pushed `c4a1f8be` addressing the K-builder SH-NB failure-path race from your previous review (line:204):

The killswitch path no longer falls through to qemu emulation when `flock -s -n` fails — that case was where a brief peer EX hold (probe / disable / EX→SH downgrade window) caused us to drop the anchor and let the peer complete its transition with qemu-arm=0. Now we use blocking `flock -s -w 30` and recheck state after acquiring SH: anchor on `state=1`, `exit_with_error` on `state=0`. Sub-millisecond peer transitions resolve well within the window; if SH stays unreachable for 30s the code still falls back without anchor but only after the timeout, not on the first probe.

The remaining findings from your prior review are deferred for now:
- **P1 line:148 rootfs-create.sh** (peer-bootstrap-in-mmdebstrap race): correct concern; needs serialization at prepare_host_binfmt_qemu_cross (host-wide EX-NB probe before any builder enters mmdebstrap, fail-fast if a concurrent native peer already won). Out of scope for the current minimal fix; will track as a follow-up.
- **P1 line:306 (cleanup ordering)**: I am still verifying call-site registration order — the empirical SIGINT test holds, but if your reading is correct (rootfs cleanup registered earlier in time → later in PREPEND list → runs after us) we need explicit descendant-kill or POSIX F_OFD on a helper fd. Will reproduce with a controlled SIGINT test before patching.
- **P2 line:24 (preserve target qemu binary)**: still treated as out-of-scope edge case (armhf image deliberately bundling qemu-user-static). Acknowledged but not in this round.

File: lib/functions/rootfs/qemu-static.sh [hunk 2]
```diff
@@ -132,6 +150,227 @@ function prepare_host_binfmt_qemu() {
...
+	esac
+
+	# Killswitch path: take SH-lock on qemu-arm so concurrent native-armhf
+	# builders detect us via EX-NB probe and refuse to switch qemu-arm off.
+	# Without this anchor an N-builder arriving mid-K-chroot would echo 0 and
+	# silently break K's qemu-arm-static routing.
+	if [[ "${killswitch}" == "yes" ]]; then
+		if [[ ! -e /proc/sys/fs/binfmt_misc/qemu-arm ]]; then
+			display_alert "Native armhf via binfmt_elf" "killswitch requested but qemu-arm not registered; cannot anchor emulation" "wrn"
+			return 1
```
[the remark this answers is not in the document]
Comment (iav) [reply lib/functions/rootfs/qemu-static.sh]: fixed

File: lib/functions/rootfs/qemu-static.sh [hunk 3]
```diff
@@ -132,6 +150,227 @@ function prepare_host_binfmt_qemu() {
...
+# (rootfs-create.sh and rootfs-image.sh), so mount_chroot's umount handlers
+# are registered after ours, sit at the head of the list, and execute before
+# us — by the time we run, the chroot is unmounted and inheriting subshells
+# have exited. Verified empirically (SIGINT mid-chroot). If a future call
+# site registers our setup AFTER mount_chroot, this invariant inverts and
+# the EX-NB probe will spuriously fail; the documented escape hatches are
+# POSIX F_SETLK on a helper fd or explicit descendant-kill.
+function trap_handler_native_armhf_restore_qemu_arm() {
+	[[ -n "${_native_armhf_lock_fd:-}" ]] || return 0
+	exec {_native_armhf_lock_fd}>&-
```
[the remark this answers is not in the document]
Comment (iav) [reply lib/functions/rootfs/qemu-static.sh:392]: fixed

**Comment by iav** [comment]
@codex review
